hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndexUtils.java

```diff
@@ -251,9 +305,14 @@ private static <R> Option<HoodieRecord<R>> mergeIncomingWithExistingRecord(
       HoodieRecord<R> existing,
       Schema writeSchema,
       HoodieWriteConfig config,
-      HoodieRecordMerger recordMerger) throws IOException {
+      HoodieRecordMerger recordMerger,
+      BaseKeyGenerator keyGenerator) throws IOException {
     Schema existingSchema = HoodieAvroUtils.addMetadataFields(new Schema.Parser().parse(config.getSchema()), config.allowOperationMetadataField());
     Schema writeSchemaWithMetaFields = HoodieAvroUtils.addMetadataFields(writeSchema, config.allowOperationMetadataField());
+    if (keyGenerator != null) {
+      return mergeIncomingWithExistingRecordWithExpressionPayload(incoming, existing, writeSchema,
+          existingSchema, writeSchemaWithMetaFields, config, recordMerger, keyGenerator);
+    }
```

Comment:
not quite sure about this dependency: keygen not null then go with expression payload.